sklearn/cluster/k_means_.py

```diff
@@ -88,6 +89,20 @@ def _k_init(X, n_clusters, x_squared_norms, random_state, n_local_trials=None):
         # that it helped.
         n_local_trials = 2 + int(np.log(n_clusters))
 
+    # Do type checks before the critical loop
+    if check_inputs:
+        X = as_float_array(X, copy=False)
```

Comment:
I didn't realize that `check_array` forced inputs into floats. I think I was following code elsewhere in the file.  I changed the `check_array` to use `dtype=sklearn.utils.FLOAT_DTYPES`.